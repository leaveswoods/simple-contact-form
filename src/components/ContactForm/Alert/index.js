import * as React from "react";
import Alert from "../../Alert";

function ContactFormAlert({ alertMessageProps, setAlertVisible }) {
  const executeOnClose = () => {
    setAlertVisible(false);
  };
  return (
    <Alert
      onClose={executeOnClose}
      message={alertMessageProps.message}
      type={alertMessageProps.type}
    />
  );
}

export default ContactFormAlert;
