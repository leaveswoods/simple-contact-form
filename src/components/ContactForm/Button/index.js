import * as React from "react";
import Button from "../../Button";

function SubmitButton({ loading, handleClickSubmit, text, style }) {
  return (
    <Button loading={loading} onClick={handleClickSubmit} style={style}>
      {text}
    </Button>
  );
}

export default SubmitButton;
