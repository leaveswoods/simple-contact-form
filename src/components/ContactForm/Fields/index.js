import * as React from "react";
import { InputField, TextField } from "../../Form/Fields/index";
import { FormItem, Group } from "../../Form/index";

function NameGroup({ required }) {
  return (
    <Group>
      <FormItem
        name="firstName"
        rules={[{ required, message: "Please enter your first name" }]}
      >
        <InputField placeholder="First Name" autoComplete="given-name" />
      </FormItem>
      <FormItem
        name="lastName"
        rules={[{ required, message: "Please enter your last name" }]}
      >
        <InputField placeholder="Last Name" autoComplete="family-name" />
      </FormItem>
    </Group>
  );
}

function Email({ required }) {
  return (
    <FormItem
      name="email"
      rules={[
        { required, message: "Please enter your email" },
        { type: "email", message: "Please enter a valid email" },
      ]}
    >
      <InputField placeholder="Email" autoComplete="email" />
    </FormItem>
  );
}

function Phone({ required }) {
  return (
    <FormItem
      name="phone"
      rules={[
        { required, message: "Please enter your phone" },
        { type: "phone", message: "Please enter a valid phone number" },
      ]}
    >
      <InputField placeholder="Phone" autoComplete="phone" />
    </FormItem>
  );
}

function Message({ required }) {
  return (
    <FormItem
      name="message"
      rules={[{ required, message: "Please leave some message for us" }]}
    >
      <TextField placeholder="Message" rows={5} />
    </FormItem>
  );
}

export { NameGroup, Email, Phone, Message };
