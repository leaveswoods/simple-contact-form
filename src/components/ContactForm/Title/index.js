import * as React from "react";

function Title({ text }) {
  return <div className="form_title mb-4 text-4xl text-center">{text}</div>;
}

export default Title;
