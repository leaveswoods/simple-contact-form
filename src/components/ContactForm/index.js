import * as React from "react";
import PropTypes from "prop-types";

import { Form, useForm } from "../Form/index";
import { NameGroup, Email, Phone, Message } from "./Fields";
import SubmitButton from "./Button";
import ContactFormAlert from "./Alert";
import Title from "./Title";
import { warning, error } from "../../utils";

function ContactFormWrapper({ children }) {
  return <div className="contact_form">{children}</div>;
}

function ContactForm({
  onSubmit,
  loading,
  alertMessageProps,
  title,
  fields,
  submitButtonProps,
}) {
  const form = useForm();
  const isShowAlertVisible = () => {
    if (
      alertMessageProps &&
      alertMessageProps.type &&
      ["success", "error"].includes(alertMessageProps.type) &&
      alertMessageProps.message
    ) {
      return true;
    }
    return false;
  };
  const [alertVisible, setAlertVisible] = React.useState(isShowAlertVisible());

  React.useEffect(() => {
    if (isShowAlertVisible()) {
      setAlertVisible(true);
    }
  }, [alertMessageProps.message, alertMessageProps.type]);

  const submitContact = async () => {
    const {
      email,
      firstName,
      lastName,
      phone,
      message,
    } = form.getFieldsValue();
    const contactDetail = {
      email,
      firstName,
      lastName,
      phone,
      message,
    };
    try {
      if (onSubmit) {
        await onSubmit(contactDetail);
        form.initFieldsValue();
      } else {
        warning(
          "You need to pass an onSubmit function in props to handle submit."
        );
      }
    } catch (e) {
      error(e);
    }
  };

  const handleClickSubmit = () => {
    const validateResult = form.validateFields();
    if (validateResult) {
      submitContact();
    }
  };
  const components = {
    nameGroup: NameGroup,
    email: Email,
    phone: Phone,
    message: Message,
  };
  return (
    <ContactFormWrapper>
      <Form
        name="contact"
        form={form}
        title={title ? <Title text={title} /> : ""}
      >
        {alertVisible && (
          <ContactFormAlert
            alertMessageProps={alertMessageProps}
            setAlertVisible={setAlertVisible}
          />
        )}
        {fields.map((field) => {
          if (typeof field === "object") {
            const Field = components[field.type];
            if (!Field) return null;
            return <Field key={field.type} required={field.required} />;
          }
          if (typeof field === "string") {
            const Field = components[field];
            if (!Field) return null;
            return <Field key={field} />;
          }
        })}
        <SubmitButton
          loading={loading}
          handleClickSubmit={handleClickSubmit}
          text={submitButtonProps.text}
          style={submitButtonProps.style}
        />
      </Form>
    </ContactFormWrapper>
  );
}
ContactForm.defaultProps = {
  loading: false,
  alertMessageProps: {
    type: "",
    message: "",
  },
  title: "",
  fields: [
    {
      type: "nameGroup",
      required: true,
    },
    {
      type: "email",
      required: true,
    },
    {
      type: "phone",
      required: false,
    },
    {
      type: "message",
      required: true,
    },
  ],
  submitButtonProps: {
    text: "Submit",
    style: {},
  },
};

ContactForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  alertMessageProps: PropTypes.shape({
    message: PropTypes.string,
    type: PropTypes.oneOf(["error", "success", ""]),
  }),
  title: PropTypes.string,
  fields: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({
        type: PropTypes.oneOf(["nameGroup", "email", "phone", "message"]),
        required: PropTypes.bool,
      }),
    ])
  ),
  submitButtonProps: PropTypes.shape({
    text: PropTypes.string,
    style: PropTypes.object,
  }),
};

export default ContactForm;
