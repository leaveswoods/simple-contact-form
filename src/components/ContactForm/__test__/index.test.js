/* eslint-disable react/jsx-props-no-spreading */
// import dependencies
import * as React from "react";
// import * as React-testing methods
import { render, fireEvent, waitFor } from "@testing-library/react";
// add custom jest matchers from jest-dom
import "@testing-library/jest-dom/extend-expect";

import ContactForm from "../index";

describe("Contact Form", () => {
  describe("Title", () => {
    it("should render title if provided title text", () => {
      const mockCallback = jest.fn();
      const { asFragment } = render(
        <ContactForm
          title="Contact us now"
          fields={["nameGroup", "email", "phone", "message"]}
          onSubmit={mockCallback}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("Fields", () => {
    it("should render all fields", () => {
      const mockCallback = jest.fn();

      const { asFragment } = render(
        <ContactForm
          fields={["nameGroup", "email", "phone", "message"]}
          onSubmit={mockCallback}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should only render specified fields", () => {
      const mockCallback = jest.fn();

      const { asFragment } = render(
        <ContactForm
          fields={["nameGroup", "email", "message"]}
          onSubmit={mockCallback}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should prompt error message if required fields are empty", async () => {
      const mockCallback = jest.fn();

      const { getAllByTestId, getByText } = render(
        <ContactForm
          fields={[
            { type: "nameGroup", required: true },
            { type: "phone", required: true },
            { type: "email", required: true },
            { type: "message", required: true },
          ]}
          onSubmit={mockCallback}
          submitButtonProps={{
            text: "Submit",
          }}
        />
      );
      fireEvent.click(getByText("Submit"));
      expect(getAllByTestId("form_field_error_message")).toMatchSnapshot();
    });

    it("should display loader when loading is true", () => {
      const mockCallback = jest.fn();
      const { getByTestId } = render(
        <ContactForm
          fields={[
            { type: "nameGroup", required: true },
            { type: "phone", required: true },
            { type: "email", required: true },
            { type: "message", required: true },
          ]}
          onSubmit={mockCallback}
          loading
        />
      );
      expect(getByTestId("btn_loader")).toMatchSnapshot();
    });
  }); // Fields

  describe("Alert", () => {
    it("should show alert if have alertMessageProps", () => {
      const mockCallback = jest.fn();
      const props = {
        fields: [
          { type: "nameGroup", required: true },
          { type: "phone", required: true },
          { type: "email", required: true },
          { type: "message", required: true },
        ],
        onSubmit: mockCallback,
        alertMessageProps: {
          type: "",
          message: "",
        },
      };
      const { queryByTestId, rerender } = render(<ContactForm {...props} />);
      expect(queryByTestId("form_alert")).not.toBeInTheDocument();

      props.alertMessageProps = {
        type: "success",
        message: "This is an alert.",
      };
      rerender(<ContactForm {...props} />);
      expect(queryByTestId("form_alert")).toBeInTheDocument();
    });

    it("should close if click close alert button", async () => {
      const mockCallback = jest.fn();
      const { getByTestId, queryByTestId } = render(
        <ContactForm
          onSubmit={mockCallback}
          alertMessageProps={{
            type: "success",
            message: "This is an alert message",
          }}
        />
      );
      fireEvent.click(getByTestId("close_btn"));
      await waitFor(() => {
        expect(queryByTestId("form_alert")).not.toBeInTheDocument();
      });
    });

    it("should not render alert if passing invalid alert type", () => {
      const mockCallback = jest.fn();
      const { queryByTestId } = render(
        <ContactForm
          onSubmit={mockCallback}
          alertMessageProps={{
            type: "dwadbwadi",
            message: "This is an alert message",
          }}
        />
      );
      expect(queryByTestId("form_alert")).not.toBeInTheDocument();
    });
  }); // Alert

  describe("Submit Button", () => {
    it("should render button with text specified", () => {
      const mockCallback = jest.fn();
      const buttonText = "Submit Button";
      const { getByText } = render(
        <ContactForm
          fields={[
            { type: "nameGroup", required: true },
            { type: "phone", required: true },
            { type: "email", required: true },
            { type: "message", required: true },
          ]}
          onSubmit={mockCallback}
          submitButtonProps={{ text: buttonText }}
        />
      );
      expect(getByText(buttonText)).toMatchSnapshot();
    });
  }); // Submit Button
}); // Contact form
