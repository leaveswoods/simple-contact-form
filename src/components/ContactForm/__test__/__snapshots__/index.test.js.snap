// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Form Fields should display loader when loading is true 1`] = `
<div
  class="btn_loader"
  data-testid="btn_loader"
/>
`;

exports[`Contact Form Fields should only render specified fields 1`] = `
<DocumentFragment>
  <div
    class="contact_form"
  >
    <div
      class="form"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: center;"
      >
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="First Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="Last Name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="input_item"
      >
        <input
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
      <div
        class="input_item"
      >
        <textarea
          placeholder="Message"
          rows="5"
        />
      </div>
      <div
        class="btn_outer"
      >
        <div
          class="btn_inner"
          role="button"
          tabindex="0"
        >
          Submit
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Contact Form Fields should prompt error message if required fields are empty 1`] = `
Array [
  <div
    class="error_message text-red-500 absolute text-sm"
    data-testid="form_field_error_message"
  >
    Please enter your first name
  </div>,
  <div
    class="error_message text-red-500 absolute text-sm"
    data-testid="form_field_error_message"
  >
    Please enter your last name
  </div>,
  <div
    class="error_message text-red-500 absolute text-sm"
    data-testid="form_field_error_message"
  >
    Please enter your phone
  </div>,
  <div
    class="error_message text-red-500 absolute text-sm"
    data-testid="form_field_error_message"
  >
    Please enter your email
  </div>,
  <div
    class="error_message text-red-500 absolute text-sm"
    data-testid="form_field_error_message"
  >
    Please leave some message for us
  </div>,
]
`;

exports[`Contact Form Fields should render all fields 1`] = `
<DocumentFragment>
  <div
    class="contact_form"
  >
    <div
      class="form"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: center;"
      >
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="First Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="Last Name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="input_item"
      >
        <input
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
      <div
        class="input_item"
      >
        <input
          placeholder="Phone"
          type="text"
          value=""
        />
      </div>
      <div
        class="input_item"
      >
        <textarea
          placeholder="Message"
          rows="5"
        />
      </div>
      <div
        class="btn_outer"
      >
        <div
          class="btn_inner"
          role="button"
          tabindex="0"
        >
          Submit
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Contact Form Submit Button should render button with text specified 1`] = `
<div
  class="btn_inner"
  role="button"
  tabindex="0"
>
  Submit Button
</div>
`;

exports[`Contact Form Title should render title if provided title text 1`] = `
<DocumentFragment>
  <div
    class="contact_form"
  >
    <div
      class="form"
    >
      <div
        class="form_title mb-4 text-4xl text-center"
      >
        Contact us now
      </div>
      <div
        style="display: flex; justify-content: space-between; align-items: center;"
      >
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="First Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          style="width: 48%;"
        >
          <div
            class="input_item"
          >
            <input
              placeholder="Last Name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="input_item"
      >
        <input
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
      <div
        class="input_item"
      >
        <input
          placeholder="Phone"
          type="text"
          value=""
        />
      </div>
      <div
        class="input_item"
      >
        <textarea
          placeholder="Message"
          rows="5"
        />
      </div>
      <div
        class="btn_outer"
      >
        <div
          class="btn_inner"
          role="button"
          tabindex="0"
        >
          Submit
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
